components/common/src/test/java/com/hotels/styx/config/validator/DocumentFormatTest.java

```diff
@@ -0,0 +1,526 @@
+                ))
+                .subSchema("PluginsList", schema(field("all", string())
+                ))
+                .rootSchema(schema(
+                        field("plugins", object("PluginsList")),
+                        field("httpPipeline", object("HttpPipeline"))
+                ))
+                .build();
+    }
+
+    @Test
+    public void validatesDiscriminatedUnions() throws Exception {
+
+        DocumentFormat validator = newDocument()
+                .subSchema("ProxyTo", schema(
+                        field("id", string()),
+                        field("destination", string())
+                ))
+                .subSchema("Redirection", schema(
+                        field("status", integer()),
+                        field("location", string())
+                ))
+                .rootSchema(schema(
+                        field("httpPipeline", object(
+                                field("type", string()),
+                                field("config", union("type"))
+                        ))
+                ))
+                .build();
+
+
+        boolean outcome1 = validator.validateObject(
+                YAML_MAPPER.readTree(
+                        "httpPipeline: \n"
+                                + "  type: 'ProxyTo'\n"
+                                + "  config:\n"
+                                + "    id: 'local-01'\n"
+                                + "    destination: 'localhost:8080'\n"
+                ));
+        assertThat(outcome1, is(true));
+
+        boolean outcome2 = validator.validateObject(
+                YAML_MAPPER.readTree(
+                        "httpPipeline: \n"
+                                + "  type: 'Redirection'\n"
+                                + "  config:\n"
+                                + "    status: 301\n"
+                                + "    location: /new/location\n"
+                ));
+        assertThat(outcome2, is(true));
+
+    }
+
+    @Test
+    public void ignoresSubobjects() throws Exception {
+        JsonNode node2 = YAML_MAPPER.readTree(
+                "parent: \n"
+                        + "  opaque: \n"
+                        + "    x: 5\n"
+                        + "    y: 6\n"
+        );
+
+        DocumentFormat validator = newDocument()
+                .rootSchema(schema(
+                        field("parent", object(
+                                field("opaque", object(opaque()))
+                        ))
+                ))
+                .build();
+
+        boolean outcome = validator.validateObject(node2);
+        assertThat(outcome, is(true));
+    }
+
+
+    @Test(expectedExceptions = SchemaValidationException.class,
+            expectedExceptionsMessageRegExp = "Schema constraint failed. At least one of \\('http', 'https'\\) must be present.")
+    public void requiresAtLeastOneOf() throws Exception {
```

Comment:
Meaningless test name.

Reply:
Please propose something better. Waiting for your feedback.

Reply:
I think the issue is that your test is checking multiple different cases at once. I would suggest at least two tests. One for the positive situation like `rejectsInvalidConfigForAtLeastOneDirective` and `acceptsValidConfigForAtLeastOneDirective`. If you prefer, you could make each of the four assertions into its own test.

Reply:
Or if you want to keep a single test, just say `validatesCorrectlyForAtLeastOneDirective`.